File: 动态规划/linear_dp/code/俄罗斯套娃信封问题.java
// https://leetcode.cn/problems/russian-doll-envelopes/

//最长上升子序列lis问题

class Solution {
    public int maxEnvelopes(int[][] envelopes) {
        if (envelopes.length == 0) {
            return 0;
        }

        int n = envelopes.length;
        Arrays.sort(envelopes, new Comparator<int[]>() {
            public int compare(int[] e1, int[] e2) {
                if (e1[0] != e2[0]) {
                    // 宽度升序
                    return e1[0] - e2[0];
                } else {
                    // 若宽度相等，则高度降序(理解高度降序)
                    return e2[1] - e1[1];
                }
            }
        });

        // 存储可被套娃的信封序列
        List<Integer> f = new ArrayList<Integer>();
        f.add(envelopes[0][1]);
        for (int i = 1; i < n; ++i) {
            int num = envelopes[i][1];
            // 直接满足条件，加到最后
            if (num > f.get(f.size() - 1)) {
                f.add(num);
            // 否则找到刚好大于等于的替换之，理解这一步，精髓！！！
            } else {
                int index = binarySearch(f, num);
                f.set(index, num);
            }
        }
        return f.size();
    }

    public int binarySearch(List<Integer> f, int target) {
        int low = 0, high = f.size() - 1;
        while (low < high) {
            int mid = (high - low) / 2 + low;
            if (f.get(mid) < target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        return low;
    }
}

File: algorithm/dp/01线性dp/code/俄罗斯套娃信封问题.cpp
class Solution {
public:
// https://leetcode.cn/problems/russian-doll-envelopes/
    int maxEnvelopes(vector<vector<int>>& envelopes) {
        sort(envelopes.begin(), envelopes.end(), cmp);

        // dp[i] 表示以第i个长方形为底，最大能堆积的高度

        vector<int> dp(envelopes.size(), 1);

        for (int i = 1; i < envelopes.size(); i++) {
            for (int j = 0; j < i; j++) {
                if (envelopes[j][0] < envelopes[i][0] && envelopes[j][1] < envelopes[i][1]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }

        int ans = 0;
        for (int i = 0; i < envelopes.size(); i++)  ans = max(ans, dp[i]);
        return ans;

    }

    // 排序算法无所谓，可以任选，只要可以体现大于就行
    static bool cmp(vector<int>&a, vector<int>&b) {
        return a[0] + a[1] < b[0] + b[1];
    }
};

// 优化

class Solution {
public:
    int maxEnvelopes(vector<vector<int>>& envelopes) {
        if (envelopes.empty()) {
            return 0;
        }
        sort(envelopes.begin(), envelopes.end(), cmp);

        // dp[i] 表示长度为i的最长上升子序列的最后一个值

        vector<int> dp = {envelopes[0][1]};

        for (int i = 1; i < envelopes.size(); i++) {
            if (envelopes[i][1] > dp.back()) {
                dp.push_back(envelopes[i][1]);
            } else {
                auto it = lower_bound(dp.begin(), dp.end(), envelopes[i][1]);
                *it = envelopes[i][1];
            }
        }
        return dp.size();

    }

    static bool cmp(vector<int>&a, vector<int>&b) {
        return a[0] < b[0] || (a[0] == b[0] && a[1] > b[1]);
    }
    
};

